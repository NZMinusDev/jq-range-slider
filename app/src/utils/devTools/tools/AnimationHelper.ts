/**
 *
 * @param drawCallBack
 * @param duration
 * @param timing - calculating the current animation state function
 * @example
 * animate(
 * (progress) => { // draw
 *    elem.style.width = progress * 100 + '%';
 *  }
 *  1000, // duration
 *  {timing: (timeFraction) => { // parabolic curve timing function
 *    return Math.pow(timeFraction, 2)
 *  }},
 * );
 *
 * // timing functions(but you can pass your own):
 * function circ(timeFraction) { return 1 - Math.sin(Math.acos(timeFraction)); }
 * function shotFromABow(x, timeFraction) { return Math.pow(timeFraction, 2) * ((x + 1) * timeFraction - x) }
 * function bounce(timeFraction) {
 *   for (let a = 0, b = 1; 1; a += b, b /= 2) {
 *     if (timeFraction >= (7 - 4 * a) / 11) {
 *       return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2)
 *     }
 *   }
 * }
 * function elastic(x, timeFraction) {
 *   return Math.pow(2, 10 * (timeFraction - 1)) * Math.cos(20 * Math.PI * x / 3 * timeFraction)
 * }
 */
export function animate(
  drawCallBack: (progress: number) => unknown,
  duration: number,
  { timing = (timeFraction: number) => timeFraction } = {}
) {
  let start = performance.now();

  requestAnimationFrame(function animate(time) {
    let timeFraction = (time - start) / duration;
    if (timeFraction > 1) timeFraction = 1;

    let progress = timing(timeFraction);

    drawCallBack(progress);

    if (timeFraction < 1) {
      requestAnimationFrame(animate);
    }
  });
}

export function power(timeFraction, {power = 2} = {}) { 
      return Math.pow(timeFraction, power)
    }
export function circ(timeFraction) {
  return 1 - Math.sin(Math.acos(timeFraction));
}
export function shotFromABow(x, timeFraction) {
  return Math.pow(timeFraction, 2) * ((x + 1) * timeFraction - x);
}
export function bounce(timeFraction) {
  for (let a = 0, b = 1; 1; a += b, b /= 2) {
    if (timeFraction >= (7 - 4 * a) / 11) {
      return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2);
    }
  }
}
export function elastic(x, timeFraction) {
  return Math.pow(2, 10 * (timeFraction - 1)) * Math.cos(((20 * Math.PI * x) / 3) * timeFraction);
}

/**
 * Jumping animation of typing text
 */
export function animateTextArea(textArea: HTMLTextAreaElement) {
  let text = textArea.value;
  let to = text.length,
    from = 0;

  animate(
    (progress) => {
      let result = (to - from) * progress + from;
      textArea.value = text.substr(0, Math.ceil(result));
    },
    5000,
    { timing: bounce }
  );

  function bounce(timeFraction: number): number {
    for (let a = 0, b = 1, result; 1; a += b, b /= 2) {
      if (timeFraction >= (7 - 4 * a) / 11) {
        return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2);
      }
    }

    return 0;
  }
}

/**
 *
 * @param timing - easeIn function
 * @returns easeOut function
 * @example
 * function bounce(timeFraction) {
 *   for (let a = 0, b = 1, result; 1; a += b, b /= 2) {
 *     if (timeFraction >= (7 - 4 * a) / 11) {
 *       return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2)
 *     }
 *   }
 * }
 *
 * let bounceEaseOut = makeEaseOut(bounce);
 */
export function makeEaseOut(timingFunction: (timeFraction: number) => number) {
  return function (timeFraction: number) {
    return 1 - timingFunction(1 - timeFraction);
  };
}
/**
 *
 * @param timing - easeIn function
 * @returns easeInOut function
 * @example
 * function bounce(timeFraction) {
 *   for (let a = 0, b = 1, result; 1; a += b, b /= 2) {
 *     if (timeFraction >= (7 - 4 * a) / 11) {
 *       return -Math.pow((11 - 6 * a - 11 * timeFraction) / 4, 2) + Math.pow(b, 2)
 *     }
 *   }
 * }
 *
 * let bounceEaseInOut = makeEaseInOut(bounce);
 */
export function makeEaseInOut(timingFunction: (timeFraction: number) => number) {
  return function (timeFraction: number) {
    if (timeFraction < 0.5) return timingFunction(2 * timeFraction) / 2;
    else return (2 - timingFunction(2 * (1 - timeFraction))) / 2;
  };
}
