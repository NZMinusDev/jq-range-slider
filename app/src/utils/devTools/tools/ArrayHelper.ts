/**
 * Adds missing and removes extra elements
 */
const fixLength = <TArray extends unknown[]>(
  arr: TArray,
  desiredLength: number,
  filler: TArray[number]
) => {
  const previousLength = arr.length;
  if (arr.length !== desiredLength) {
    // eslint-disable-next-line no-param-reassign
    arr.length = desiredLength;
    // eslint-disable-next-line no-param-reassign
    arr = arr.fill(filler, previousLength);
  }
};

// eslint-disable-next-line import/prefer-default-export
export { fixLength };
