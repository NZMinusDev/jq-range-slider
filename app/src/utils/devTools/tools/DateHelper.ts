/**
 * @returns milliseconds
 */
export function diffDate(date1: Date, date2: Date) {
  return date2.getTime() - date1.getTime();
}
