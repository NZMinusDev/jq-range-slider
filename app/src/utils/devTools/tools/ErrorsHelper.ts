export class MyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;
  }
}

export class ErrorWrapper extends Error {
  public cause: Error;

  constructor(message: string, cause: Error) {
    super(message);
    this.cause = cause;
    this.name = "ReadError";
  }
}

/**
 * 
 * @example
 * type NetworkLoadingState = { state: "loading"; };
 * type NetworkFailedState = { state: "failed"; code: number; };
 * type NetworkSuccessState = {
 *   state: "success";
 *   response: {
 *    title: string;
 *    duration: number;
 *    summary: string;
 *   };
 * };
 * 
 * // Create a type which represents only one of the above types
 * // but you aren't sure which it is yet.
 * type NetworkState =
 *   | NetworkLoadingState
 *   | NetworkFailedState
 *   | NetworkSuccessState;
 * 
 * function logger(state: NetworkState): string {
 *   // Right now TypeScript does not know which of the three
 *   // potential types state could be.
 *   // Trying to access a property which isn't shared
 *   // across all types will raise an error
 *   state.code;
 *   // By switching on state, TypeScript can narrow the union
 *   // down in code flow analysis
 *   switch (state.state) {
 *     case "loading":
 *       return "Downloading...";
 *     case "failed":
 *       // The type must be NetworkFailedState here,
 *       // so accessing the `code` field is safe
 *       return `Error ${state.code} downloading`;
 *     case "success":
 *       return `Downloaded ${state.response.title} - ${state.response.summary}`;
 *     default:
 *       return assertNever(s);
 *   }
 * }
 */
export function assertNever(valueToCheck: never): never {
  throw new Error("Unexpected object: " + valueToCheck);
}