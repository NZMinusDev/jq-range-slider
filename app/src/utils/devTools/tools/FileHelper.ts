export function createDataURL(
  blobParts: BlobPart[],
  options: BlobPropertyBag = { type: "text/plain" }
) {
  let blob = new Blob(blobParts, options);
  return URL.createObjectURL(blob);
}

export async function createBase64(
  blobParts: BlobPart[],
  options: BlobPropertyBag = { type: "text/plain" }
) {
  let blob = new Blob(blobParts, options);

  let fileReader = new FileReader();
  fileReader.readAsDataURL(blob);

  return await new Promise<string>((resolve) => {
    fileReader.onload = function () {
      resolve(fileReader.result as string);
    };
  });
}

export async function blobToArrayBuffer(blob: Blob) {
  let fileReader = new FileReader();
  fileReader.readAsArrayBuffer(blob);

  return await new Promise<ArrayBuffer>((resolve) => {
    fileReader.onload = function () {
      resolve(fileReader.result as ArrayBuffer);
    };
  });
}

export function imgToCanvas(
  img: HTMLImageElement,
  imgProcessing = function (canvas: HTMLCanvasElement) {}
) {
  let canvas = document.createElement("canvas");
  canvas.width = img.clientWidth;
  canvas.height = img.clientHeight;

  let context = canvas.getContext("2d");
  context?.drawImage(img, 0, 0);

  imgProcessing(canvas);

  return canvas;
}
export async function imgToBlob(
  img: HTMLImageElement,
  imgProcessing = function (canvas: HTMLCanvasElement) {}
) {
  const canvas = imgToCanvas(img, imgProcessing);

  return await new Promise<Blob>((resolve) => canvas.toBlob(resolve as BlobCallback, "image/png"));
}
