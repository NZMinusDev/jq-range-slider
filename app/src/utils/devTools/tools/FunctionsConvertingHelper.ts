import { GenericFunc } from "./TypingHelper";

/**
 * Currying of function
 * @param func
 * @example
 * function sum(a, b, c) {
 *  return a + b + c;
 *}
 *
 *let curriedSum = curry(sum);
 *
 *alert( curriedSum(1, 2, 3) ); // 6
 *alert( curriedSum(1)(2,3) ); // 6
 *alert( curriedSum(1)(2)(3) ); // 6
 */
export function curry(func: (...args: unknown[]) => unknown) {
  return function curried(...args: unknown[]) {
    if (args.length >= func.length) {
      return func.apply(this, args);
    } else {
      return function (...args: unknown[]) {
        return curried.apply(this, args.concat(args));
      };
    }
  };
}

/**
 * Clone function
 * @param func function to clone
 * @returns clone of function
 * @example
 * const test1 =  function(a,b,c) {
 *   return a+b+c;
 * };
 *
 * console.log(test1 === clone(test1)); // false
 * console.log(test1(1,1,1) === clone(test1)(1,1,1)); // true
 * console.log(clone(test1)(1,1,1)); // 3
 */
export function cloneFunction<TFuncArgs extends unknown[], TFuncReturn>(
  func: GenericFunc<TFuncArgs, TFuncReturn> | any
) {
  const that = func;
  const temp = function () {
    return that.apply(func, (arguments as unknown) as TFuncArgs);
  };

  for (let key in func) {
    if (func.hasOwnProperty(key)) {
      temp[key] = func[key];
    }
  }

  Object.defineProperty(temp, "name", { value: func.name });

  return temp;
}
