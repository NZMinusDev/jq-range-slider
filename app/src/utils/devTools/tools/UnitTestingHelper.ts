import cloneDeep from 'lodash-es/cloneDeep';
import isPlainObject from 'lodash-es/isPlainObject';

import { OptionalTupleValues, RequiredTupleValues } from './TypingHelper';
import { eachDeep, resolveLongBracketNotation } from './ObjectHelper';
import { isReferenceType } from './TypeOf';
import { renderMVPView } from './PluginCreationHelper';

type InstancePropsExpecter<TArgs extends unknown[], TInstance> = (parts: {
  passedArgs: TArgs;
  instance: TInstance;
}) => void;

const expectValidInstanceProps = <
  TInstance,
  TCreatorArgs extends unknown[],
  TMethodArgs extends unknown[]
>(
  passedArgs: TCreatorArgs | TMethodArgs,
  expecter: InstancePropsExpecter<TCreatorArgs | TMethodArgs, TInstance>,
  instance: TInstance
): void => {
  test("instance's properties should be defined", () => {
    eachDeep(instance, ({ value }) => {
      expect(value).toBeDefined();
    });
  });

  test("instance's properties should be bounded", () => {
    expecter({
      passedArgs,
      instance,
    });
  });
};

/**
 * Testing(expect non undefined bounded(by instancePropsExpecter) properties) of arguments of constructor by creating the instance with argsOfCreator
 * @param Creator A Class or Function
 * @param argsOfCreator default arguments for constructor
 * @param instancePropsExpecter callback with expect calls
 */
const testDefaultOptions = <
  TCreatorArgs extends unknown[],
  TInstance,
  TCreator extends new (...args: TCreatorArgs) => TInstance
>(
  Creator: TCreator,
  argsOfCreator: TCreatorArgs,
  instancePropsExpecter: InstancePropsExpecter<TCreatorArgs, TInstance>
) => {
  describe('DEFAULT_OPTIONS', () => {
    expectValidInstanceProps(argsOfCreator, instancePropsExpecter, new Creator(...argsOfCreator));
  });
};

type DifferentArguments<TArgs extends unknown[]> = {
  validRequiredArguments?: RequiredTupleValues<TArgs>[];
  invalidRequiredArguments?: RequiredTupleValues<TArgs>[];
  invalidOptionalArguments?: OptionalTupleValues<TArgs>[];
  partialOptionalArguments?: OptionalTupleValues<TArgs>[];
  fullOptionalArguments?: OptionalTupleValues<TArgs>[];
};

type InstanceMethodExpecter<TArgs extends unknown[], TInstance> = (parts: {
  mock: jest.Mock;
  passedArgs: TArgs;
  instance: TInstance;
}) => void;

type PathToPropInInstance = string;
type PropToGet = PathToPropInInstance;
type PropsToSet<TArgValue> = Map<PathToPropInInstance, TArgValue>;
type ResetPropsTo<TDefaultValue> = Map<PathToPropInInstance, TDefaultValue>;
type MethodOfInstanceToRun<
  TMethodArgs extends unknown[],
  TMethod extends (...args: TMethodArgs) => unknown,
  TInstance
> = {
  methodReference: TMethod | string;
  argsToPass: TMethodArgs;
  returns?: PropToGet;
  expecter: InstanceMethodExpecter<TMethodArgs, TInstance>;
};
type MethodOfInstanceToTest<
  TMethodArgs extends unknown[],
  TMethod extends (...args: TMethodArgs) => unknown,
  TInstance
> = {
  methodReference: TMethod | string;
  expecter: InstanceMethodExpecter<TMethodArgs, TInstance>;
  differentArguments?: DifferentArguments<TMethodArgs>;
  returns?: PropToGet;
};

const eachOfDifferentArguments = <TArgs extends unknown[]>(
  differentArguments: DifferentArguments<TArgs>,
  callbackEach: (
    key: string,
    concreteDifferentArgs: TArgs[],
    index: number,
    isRequiredArgs: boolean,
    isValidArgs: boolean
  ) => void
) => {
  let counter = 0;

  Object.entries(differentArguments).forEach(([key, arrayOfConcreteDifferentArgs]) => {
    (arrayOfConcreteDifferentArgs as TArgs[][] | undefined)?.forEach(
      (concreteDifferentArgs, index) => {
        callbackEach(
          key,
          concreteDifferentArgs,
          index,
          key === 'validRequiredArguments' || key === 'invalidRequiredArguments',
          key === 'validRequiredArguments' ||
            key === 'partialOptionalArguments' ||
            key === 'fullOptionalArguments'
        );

        counter += 1;
      }
    );
  });

  return counter;
};

const expectPassedArgsAreImmutable = <TArgs extends unknown[]>(
  passedArgs: TArgs,
  immutableCopyOfPassedArgs: TArgs
): void => {
  passedArgs.forEach((passedArgValue, index) => {
    if (isReferenceType(passedArgValue)) {
      test(`${index + 1} argument should be immutable`, () => {
        expect(passedArgValue).not.toBe(immutableCopyOfPassedArgs[index]);
        expect(passedArgValue).toStrictEqual(immutableCopyOfPassedArgs[index]);
      });
    }
  });
};

const expectReturnedAreImmutable = <TInstance>(
  returned: unknown,
  returns: PropToGet,
  instance: TInstance
): void => {
  const propOfInstanceValue = resolveLongBracketNotation(returns, instance);
  if (isReferenceType(propOfInstanceValue)) {
    test(`returned referenced property should be cloned deep`, () => {
      expect(returned).not.toBe(propOfInstanceValue);
      expect(returned).toStrictEqual(propOfInstanceValue);

      eachDeep(returned, ({ value, path }) => {
        if (isReferenceType(value) && typeof value !== 'function') {
          expect(resolveLongBracketNotation(`${returns}.${path}`, instance)).not.toBe(value);
          expect(resolveLongBracketNotation(`${returns}.${path}`, instance)).toMatchObject(
            value as any
          );
        }
      });
    });
  }
};

const expectMethodOfInstance = <TArgs extends unknown[], TInstance>(
  describeSentence: string,
  mock: jest.Mock,
  passedArgs: TArgs,
  expecter: InstanceMethodExpecter<TArgs, TInstance>,
  instance: TInstance
) => {
  test(describeSentence, () => {
    expecter({ mock, passedArgs, instance });
  });
};

const expectPropsAreApplied = <TInstance>(
  propsToSet: PropsToSet<unknown>,
  instance: TInstance
): void => {
  test("correct user's arguments should be applied", () => {
    propsToSet.forEach((argValue, pathToPropInInstance) => {
      if (argValue !== undefined) {
        if (isPlainObject(argValue)) {
          expect(resolveLongBracketNotation(`${pathToPropInInstance}`, instance)).toMatchObject(
            argValue as any
          );
        } else {
          expect(resolveLongBracketNotation(`${pathToPropInInstance}`, instance)).toEqual(argValue);
        }
      }
    });
  });

  propsToSet.forEach((argValue, pathToPropInInstance) => {
    if (isReferenceType(argValue)) {
      test(`passed referenced argument setting ${pathToPropInInstance} instance's property should be cloned deep`, () => {
        expect(resolveLongBracketNotation(`${pathToPropInInstance}`, instance)).not.toBe(argValue);
        expect(resolveLongBracketNotation(`${pathToPropInInstance}`, instance)).toMatchObject(
          argValue as any
        );

        eachDeep(argValue, ({ value, path }) => {
          if (isReferenceType(value) && typeof value !== 'function') {
            expect(
              resolveLongBracketNotation(`${pathToPropInInstance}.${path}`, instance)
            ).not.toBe(value);
            expect(
              resolveLongBracketNotation(`${pathToPropInInstance}.${path}`, instance)
            ).toMatchObject(value as any);
          }
        });
      });
    }
  });
};

const testInstance = <
  TCreatorArgs extends unknown[],
  TMethodArgs extends unknown[],
  TMethod extends (...args: TMethodArgs) => unknown,
  TInstance,
  TCreator extends new (...args: TCreatorArgs) => TInstance
>({
  describeSentence,
  Creator,
  constructorArgs,
  instancePropsExpecter,
  methodOfInstanceToRun,
  propsToSet,
}: {
  describeSentence: string;
  Creator: TCreator;
  constructorArgs: TCreatorArgs;
  instancePropsExpecter: InstancePropsExpecter<TCreatorArgs | TMethodArgs, TInstance>;
  methodOfInstanceToRun?: MethodOfInstanceToRun<TMethodArgs, TMethod, TInstance>;
  propsToSet?: PropsToSet<unknown>;
}) => {
  describe(describeSentence, () => {
    const copyOfConstructorArgs = cloneDeep(constructorArgs);
    const instance: TInstance = new Creator(...constructorArgs);
    let passedArgs: TCreatorArgs | TMethodArgs = constructorArgs;

    if (!methodOfInstanceToRun) {
      expectPassedArgsAreImmutable(constructorArgs, copyOfConstructorArgs);
    } else {
      const methodOfInstanceMock = jest.fn(
        instance[
          typeof methodOfInstanceToRun.methodReference === 'string'
            ? methodOfInstanceToRun.methodReference
            : methodOfInstanceToRun.methodReference.name
        ]
      );
      passedArgs = methodOfInstanceToRun.argsToPass;
      const copyOfPassedArgs = cloneDeep(passedArgs);

      const mockReturned = methodOfInstanceMock.apply(instance, passedArgs);

      expectPassedArgsAreImmutable(passedArgs, copyOfPassedArgs);
      if (methodOfInstanceToRun.returns) {
        expectReturnedAreImmutable(mockReturned, methodOfInstanceToRun.returns, instance);
      }

      expectMethodOfInstance(
        `method ${
          typeof methodOfInstanceToRun.methodReference === 'string'
            ? methodOfInstanceToRun.methodReference
            : methodOfInstanceToRun.methodReference.name
        } should pass methodOfInstanceToTest.expecter checks (possibly decorated)`,
        methodOfInstanceMock,
        passedArgs as TMethodArgs,
        methodOfInstanceToRun.expecter,
        instance
      );
    }

    if (propsToSet) {
      expectPropsAreApplied(propsToSet, instance);
    }

    expectValidInstanceProps(passedArgs, instancePropsExpecter, instance);
  });
};

const runMethodOfInstanceWithDifferentArguments = <
  TCreatorArgs extends unknown[],
  TMethodArgs extends unknown[],
  TMethod extends (...args: TMethodArgs) => unknown,
  TInstance,
  TCreator extends new (...args: TCreatorArgs) => TInstance
>({
  Creator,
  differentConstructorArgs,
  instancePropsExpecter,
  methodOfInstanceToTest,
  propsToSet,
}: {
  Creator: TCreator;
  differentConstructorArgs: DifferentArguments<TCreatorArgs>;
  instancePropsExpecter: InstancePropsExpecter<TCreatorArgs | TMethodArgs, TInstance>;
  methodOfInstanceToTest?: MethodOfInstanceToTest<TMethodArgs, TMethod, TInstance>;
  propsToSet?: PropsToSet<number | string>;
  // eslint-disable-next-line sonarjs/cognitive-complexity
}) => {
  const requiredConstructorArgs = (differentConstructorArgs.validRequiredArguments
    ? differentConstructorArgs.validRequiredArguments[0]
    : []) as TCreatorArgs;
  const requiredMethodArgs = (methodOfInstanceToTest?.differentArguments?.validRequiredArguments
    ? methodOfInstanceToTest.differentArguments.validRequiredArguments[0]
    : []) as TMethodArgs;

  let constructorArgs = requiredConstructorArgs;
  let differentArguments: DifferentArguments<TCreatorArgs | TMethodArgs> = differentConstructorArgs;
  let methodOfInstanceToRun: MethodOfInstanceToRun<TMethodArgs, TMethod, TInstance> | undefined;

  describe(`${
    typeof methodOfInstanceToTest?.methodReference === 'string'
      ? methodOfInstanceToTest?.methodReference
      : methodOfInstanceToTest?.methodReference.name || 'init'
  }`, () => {
    if (methodOfInstanceToTest) {
      methodOfInstanceToRun = {
        methodReference: methodOfInstanceToTest.methodReference,
        argsToPass: requiredMethodArgs,
        returns: methodOfInstanceToTest.returns,
        expecter: methodOfInstanceToTest.expecter,
      };

      differentArguments = methodOfInstanceToTest.differentArguments || {};
    }

    let propsToSetWithValues: PropsToSet<unknown> | undefined;

    const amountOfDifferentArgs = eachOfDifferentArguments(
      differentArguments,
      (key, concreteDifferentArgs, index, isRequiredArgs, isValidArgs) => {
        if (methodOfInstanceToRun) {
          methodOfInstanceToRun.argsToPass = isRequiredArgs
            ? (concreteDifferentArgs as TMethodArgs)
            : (requiredMethodArgs.concat(concreteDifferentArgs) as TMethodArgs);
        } else {
          constructorArgs = isRequiredArgs
            ? (concreteDifferentArgs as TCreatorArgs)
            : (requiredConstructorArgs.concat(concreteDifferentArgs) as TCreatorArgs);
        }

        if (propsToSet !== undefined && isValidArgs) {
          propsToSetWithValues = new Map();

          propsToSet.forEach((referenceOfArgToSet, pathToPropInInstance) => {
            propsToSetWithValues?.set(
              pathToPropInInstance,
              resolveLongBracketNotation(
                `${referenceOfArgToSet}`,
                methodOfInstanceToRun?.argsToPass || constructorArgs
              )
            );
          });
        } else {
          propsToSetWithValues = undefined;
        }

        testInstance({
          describeSentence: `with ${index + 1} ${key} arguments`,
          Creator,
          constructorArgs,
          instancePropsExpecter,
          methodOfInstanceToRun,
          propsToSet: propsToSetWithValues,
        });
      }
    );

    if (amountOfDifferentArgs === 0) {
      testInstance({
        describeSentence: `without arguments`,
        Creator,
        constructorArgs,
        instancePropsExpecter,
        methodOfInstanceToRun,
        propsToSet: propsToSetWithValues,
      });
    }
  });
};

/**
 * Testing(expect non undefined, bounded(by instancePropsExpecter), setting up(by propsToSet), immutable(deep cloned) properties of instance after init and immutable passed args)
 * @param Creator A Class or Function
 * @param differentConstructorArgs different arguments for constructor: validRequiredArguments - the first will be used as shim for optional args, invalidRequiredArguments, invalidOptionalArguments, partialOptionalArguments - part of valid optional args, fullOptionalArguments - full valid optional args
 * @param instancePropsExpecter callback with expect calls
 * @param propsToSet association of valid constructor's args with properties of instance, for example: new Map().set("_options", 0).set("taxes.3", 1).set("payout","2.doc.payout") - It means the first arg set this._options property and the second one set this.taxes[3] property, the third one set this.payout as doc.payout property of 2 argument
 */
const testInit = <
  TCreatorArgs extends unknown[],
  TInstance,
  TCreator extends new (...args: TCreatorArgs) => TInstance
>({
  Creator,
  differentConstructorArgs,
  instancePropsExpecter,
  propsToSet,
}: {
  Creator: TCreator;
  differentConstructorArgs: DifferentArguments<TCreatorArgs>;
  instancePropsExpecter: InstancePropsExpecter<TCreatorArgs, TInstance>;
  propsToSet?: PropsToSet<number | string>;
}) => {
  runMethodOfInstanceWithDifferentArguments({
    Creator,
    differentConstructorArgs,
    instancePropsExpecter,
    propsToSet,
  });
};

/**
 * Testing(expect non undefined bounded(by instancePropsExpecter) properties of instance. Calls toHaveReturnedWith expecter for the method, returned value is expected be cloned)
 * @param Creator A Class or Function
 * @param constructorArgs arguments to pass to constructor
 * @param instancePropsExpecter callback with expect calls for instance state
 * @param methodOfInstanceToTest data about tested method
 */
const testGetter = <
  TCreatorArgs extends unknown[],
  TMethod extends (args: any) => any,
  TInstance,
  TCreator extends new (...args: TCreatorArgs) => TInstance
>({
  Creator,
  constructorArgs,
  instancePropsExpecter,
  methodOfInstanceToTest,
}: {
  Creator: TCreator;
  constructorArgs: TCreatorArgs;
  instancePropsExpecter: InstancePropsExpecter<TCreatorArgs, TInstance>;
  methodOfInstanceToTest: MethodOfInstanceToTest<Parameters<TMethod>, TMethod, TInstance>;
}) => {
  const makeDefaultGetterExpecter = (
    expecter: InstanceMethodExpecter<Parameters<TMethod>, TInstance>
  ): InstanceMethodExpecter<Parameters<TMethod>, TInstance> => ({ mock, passedArgs, instance }) => {
    if (methodOfInstanceToTest.returns) {
      expect(mock).toHaveReturnedWith(
        resolveLongBracketNotation(methodOfInstanceToTest.returns, instance)
      );
    }

    expecter({ mock, passedArgs, instance });
  };

  // eslint-disable-next-line no-param-reassign
  methodOfInstanceToTest.expecter = makeDefaultGetterExpecter(methodOfInstanceToTest.expecter);

  runMethodOfInstanceWithDifferentArguments({
    Creator,
    differentConstructorArgs: { validRequiredArguments: [constructorArgs] } as DifferentArguments<
      TCreatorArgs
    >,
    instancePropsExpecter: instancePropsExpecter as InstancePropsExpecter<
      TCreatorArgs | Parameters<TMethod>,
      TInstance
    >,
    methodOfInstanceToTest,
  });
};

/**
 * Testing(expect non undefined bounded(by instancePropsExpecter) properties of instance. Also expects the method sets targeted property, resets property to default if undefined is passed and returns this. Expects passed arguments are immutable and deep cloned).
 * @param Creator A Class or Function
 * @param constructorArgs arguments to pass to constructor, should be full(with optional arguments) if you use resetPropsTo with the same default options as in constructor for reset testing purpose
 * @param instancePropsExpecter callback with expect calls for instance state
 * @param methodOfInstanceToTest data about tested method
 * @param propsToSet association of valid method's args with properties of instance, for example: new Map().set("_options", 0).set("taxes.3", 1).set("payout","2.doc.payout") - It means the first arg set this._options property and the second one set this.taxes[3] property, the third one set this.payout as doc.payout property of 2 argument
 * @param resetPropsTo association of properties of instance with reset to values, for example: new Map().set("_options", DEFAULT_OPTIONS) - It means this._options should be DEFAULT_OPTIONS after resetting
 */
const testSetter = <
  TCreatorArgs extends unknown[],
  TMethod extends (...args: any) => any,
  TInstance,
  TCreator extends new (...args: TCreatorArgs) => TInstance
>({
  Creator,
  constructorArgs,
  instancePropsExpecter,
  methodOfInstanceToTest,
  propsToSet,
  resetPropsTo,
}: {
  Creator: TCreator;
  constructorArgs: TCreatorArgs;
  instancePropsExpecter: InstancePropsExpecter<TCreatorArgs, TInstance>;
  methodOfInstanceToTest: MethodOfInstanceToTest<Parameters<TMethod>, TMethod, TInstance>;
  propsToSet: PropsToSet<number | string>;
  resetPropsTo: ResetPropsTo<unknown>;
}) => {
  const argsToReset =
    methodOfInstanceToTest.differentArguments?.validRequiredArguments !== undefined
      ? methodOfInstanceToTest.differentArguments.validRequiredArguments[0]
      : [];

  const makeDefaultSetterExpecter = (
    expecter: InstanceMethodExpecter<Parameters<TMethod>, TInstance>
  ): InstanceMethodExpecter<Parameters<TMethod>, TInstance> => ({ mock, passedArgs, instance }) => {
    expect(mock).toHaveReturnedWith(instance);

    // reset when undefined is passed
    if (Object.is(passedArgs, argsToReset)) {
      resetPropsTo.forEach((defaultValue, pathToPropInInstance) => {
        expect(instance[pathToPropInInstance]).toEqual(defaultValue);
      });
    }

    expecter({ mock, passedArgs, instance });
  };

  // eslint-disable-next-line no-param-reassign
  methodOfInstanceToTest.expecter = makeDefaultSetterExpecter(methodOfInstanceToTest.expecter);
  runMethodOfInstanceWithDifferentArguments({
    Creator,
    differentConstructorArgs: { validRequiredArguments: [constructorArgs] } as DifferentArguments<
      TCreatorArgs
    >,
    instancePropsExpecter: instancePropsExpecter as InstancePropsExpecter<
      TCreatorArgs | Parameters<TMethod>,
      TInstance
    >,
    methodOfInstanceToTest,
    propsToSet,
  });

  // reset when undefined is passed
  const methodOfInstanceToRun = {
    methodReference: methodOfInstanceToTest.methodReference,
    argsToPass: (argsToReset as unknown) as Parameters<TMethod>,
    expecter: methodOfInstanceToTest.expecter,
  } as MethodOfInstanceToRun<Parameters<TMethod>, TMethod, TInstance>;
  describe(`${
    typeof methodOfInstanceToTest.methodReference === 'string'
      ? methodOfInstanceToTest.methodReference
      : methodOfInstanceToTest.methodReference.name
  }`, () => {
    testInstance({
      describeSentence: 'reset (undefined is passed)',
      Creator,
      constructorArgs,
      instancePropsExpecter: instancePropsExpecter as InstancePropsExpecter<
        TCreatorArgs | Parameters<TMethod>,
        TInstance
      >,
      methodOfInstanceToRun,
    });
  });
};

// FIXME: TTemplateArgs typing
/**
 * It runs tests with corresponding constructor and its template arguments and provides rendered container, also uses toMatchSnapshot
 * @param Creator A Class or Function
 * @param constructorsArgs arguments of constructors
 * @param templatesArgs arguments of templates
 * @param callbacksWithTest callback which runs test
 */
const testDOM = <TCreator extends new (...args: any) => InstanceType<TCreator>>({
  Creator,
  constructorsArgs,
  callbacksWithTest,
}: {
  Creator: TCreator;
  constructorsArgs: ConstructorParameters<TCreator>[];
  templatesArgs: any[];
  callbacksWithTest: (({
    container,
    instance,
  }: {
    container: DocumentFragment;
    instance: InstanceType<TCreator>;
  }) => void)[];
}) => {
  describe('DOM manipulation', () => {
    let container: DocumentFragment;
    let instance: InstanceType<TCreator>;

    callbacksWithTest.forEach((testCallback, index) => {
      container = new DocumentFragment();
      instance = renderMVPView(
        Creator,
        constructorsArgs[index] === undefined ? [] : (constructorsArgs[index] as any),
        container
      );

      testCallback({ container, instance });
    });

    container = new DocumentFragment();
    renderMVPView(Creator, [], container);
    test('renders correctly', () => {
      expect(container).toMatchSnapshot();
    });
  });
};

export {
  InstancePropsExpecter,
  testDefaultOptions,
  DifferentArguments,
  InstanceMethodExpecter,
  testInit,
  testGetter,
  testSetter,
  testDOM,
};
