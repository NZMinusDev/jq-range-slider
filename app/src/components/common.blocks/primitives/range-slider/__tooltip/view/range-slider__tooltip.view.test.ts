import RangeSliderTooltipView, {
  DEFAULT_OPTIONS,
  TooltipOptions,
} from "./range-slider__tooltip.view";

import { testInit } from "@utils/devTools/tools/UnitTestingHelper";

testInit(RangeSliderTooltipView, DEFAULT_OPTIONS, {
  partialOptions: [{ isHidden: true }],
  fullOptions: [
    {
      formatter: {
        to: (value: number) => `${value.toString()}$`,
        from: (value: string) => +parseFloat(value).toFixed(4),
      },
      isHidden: true,
    },
  ],
  expectValidProperties({ viewOptions }) {
    expect(jest.fn(viewOptions.formatter?.from)("12.345")).toHaveReturned();
    expect(jest.fn(viewOptions.formatter?.to)(12.345)).toHaveReturned();
  },
});
