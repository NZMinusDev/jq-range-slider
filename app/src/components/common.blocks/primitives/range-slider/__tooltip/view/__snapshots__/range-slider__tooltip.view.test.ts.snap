// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM manipulation renders correctly 1`] = `
<DocumentFragment>
  <!---->
  <div
    class="range-slider__tooltip"
    style=""
  >
    
    
    -1.00
    
  
  </div>
  <!---->
</DocumentFragment>
`;
